src/test/java/org/kiwiproject/registry/consul/server/ConsulRegistryServiceIntegrationTest.java

```diff
@@ -0,0 +1,199 @@
+
+            assertThatThrownBy(() -> service.register(ServiceInstance.builder().build()))
+                    .isInstanceOf(IllegalStateException.class)
+                    .hasMessageContaining("Cannot register. Already managing a registered instance");
+        }
+
+        @Test
+        void shouldRegisterAndReturnRegisteredInstance() {
+            var now = Instant.now();
+            when(environment.currentInstant()).thenReturn(now);
+
+            var serviceInstance = ServiceInstance.fromServiceInfo(ServiceInfoHelper.buildTestServiceInfo())
+                    .withStatus(ServiceInstance.Status.UP);
+
+            var registeredInstance = service.register(serviceInstance);
+
+            assertThat(registeredInstance).isNotSameAs(serviceInstance);
+
+            var storedInstance = service.registeredService.get();
+            assertThat(storedInstance).isNotNull();
+            assertThat(storedInstance.getInstanceId()).isNotBlank();
+
+            assertThat(consul.catalogClient().getService(serviceInstance.getServiceName()).getResponse()).isNotEmpty();
+        }
+
+        @Test
+        void shouldRetryRegisterAndThrowExceptionIfAllTriesExpire() {
+            var badConsul = mock(Consul.class);
+            var serviceToFail = new ConsulRegistryService(badConsul, new ConsulRegistrationConfig(), environment);
+
+            doThrow(new RuntimeException("Oops")).when(badConsul).agentClient();
+
+            var now = Instant.now();
+            when(environment.currentInstant()).thenReturn(now);
+
+            var serviceInstance = ServiceInstance.fromServiceInfo(ServiceInfoHelper.buildTestServiceInfo())
+                    .withStatus(ServiceInstance.Status.UP);
+
+            assertThatThrownBy(() -> serviceToFail.register(serviceInstance))
+                    .isInstanceOf(RegistrationException.class)
+                    .hasMessageStartingWith("Unable to register service " + serviceInstance.getServiceName() + ", id ")
+                    .hasMessageEndingWith(" with Consul after 60 attempts");
+
+            assertThat(serviceToFail.registeredService.get()).isNull();
+        }
+
+        @Test
+        void shouldRegisterAndReturnRegisteredInstance_OverridingDomainIfConfiguredToDoSo() {
+            var now = Instant.now();
+            when(environment.currentInstant()).thenReturn(now);
+
+            var serviceInstance = ServiceInstance.fromServiceInfo(ServiceInfoHelper.buildTestServiceinfoWithHostName("example.com"))
+                    .withStatus(ServiceInstance.Status.UP);
+            config.setDomainOverride("test");
+
+            service.register(serviceInstance);
+
+            var services = consul.catalogClient().getService(serviceInstance.getServiceName()).getResponse();
+
+            var hostName = services.stream().map(CatalogService::getServiceAddress).findFirst();
+            assertThat(hostName).isPresent().get().isEqualTo("example.test");
```

Comment:
For this one, I know you can use the `OptionalAssert` methods, e.g. `assertThat(hostName).hasValue("example.com");`